### ISSUE TITLE
systemd.mount-extra= does not work

### ISSUE THREAD

**Issue by yuwata** [systemd/systemd#28516]
### systemd version the issue has been seen with

254~rc3

### Used distribution

Fedora rawhide

### CPU architectures issue was seen on

x86_64

### Unexpected behaviour you saw

Originally posted by @bmr-cymru in https://github.com/systemd/systemd/issues/27260#issuecomment-1649986565.

I'm trying to test the new `systemd.mount-extra` in Fedora Rawhide with systemd-254~rc3-1.fc39.x86_64 and LVM2 logical volumes and I'm running into some problems.

Firstly, when using command line mount units and `nofstab` nothing pulls in `systemd-remount-fs.service`, which leaves the root fs mounted read-only: I think that unit should be pulled in if command line or credentials fstab entries are present. Other than that the basic case of booting a system with the fstab entries on the kernel command line seems to work as expected:

```
root=/dev/fedora/root ro rd.lvm.lv=fedora/root rd.lvm.lv=fedora/var rd.lvm.lv=fedora/home fstab=no systemd.mount-extra=/dev/fedora/root:/:ext4:rw systemd.mount-extra=/dev/vda2:/boot:ext4:defaults systemd.mount-extra=/dev/fedora/var:/var:ext4:defaults systemd.mount-extra=/dev/fedora/home:/home:ext4:defaults
```

The system boots normally and file systems are mounted as expected:

```
# grep fedora /proc/mounts 
/dev/mapper/fedora-root / ext4 ro,seclabel,relatime 0 0
/dev/mapper/fedora-home /home ext4 rw,seclabel,relatime 0 0
/dev/mapper/fedora-var /var ext4 rw,seclabel,relatime 0 0
```

Attempting to boot into a snapshot of the system state fails unpredictably: in some cases the system appears to boot normally but systemd attempts to unmount the command line mounts after switching root. In other cases the devices for the command line mount units time out. I'm struggling to debug this: the generated unit files seem sane and I can't figure out why the umounts and timeouts are occurring.

The snapshot command line looks like this:

```
root=/dev/fedora/root-snap ro rd.lvm.lv=fedora/root-snap rd.lvm.lv=fedora/var-snap rd.lvm.lv=fedora/home-snap fstab=no systemd.mount-extra=/dev/fedora/root-snap:/:ext4:rw systemd.mount-extra=/dev/vda2:/boot:ext4:defaults systemd.mount-extra=/dev/fedora/var-snap:/var:ext4:defaults systemd.mount-extra=/dev/fedora/home-snap:/home:ext4:defaults
```

In the case that file systems are unmounted the logs look like this:

```
Jul 25 14:49:36 fedora systemd[1]: Mounting sys-kernel-config.mount - Kernel Configuration File System...
Jul 25 14:49:36 fedora systemd[1]: Mounted sys-kernel-config.mount - Kernel Configuration File System.
Jul 25 14:49:36 fedora systemd[1]: Mounting boot.mount - /boot...
Jul 25 14:49:36 fedora kernel: EXT4-fs (vda2): recovery complete
Jul 25 14:49:36 fedora kernel: EXT4-fs (vda2): mounted filesystem cbb2caf6-429f-4a6a-acc4-9a20564c3f58 r/w with ordered data mode. Quota mode: none.
Jul 25 14:49:36 fedora systemd[1]: Mounted boot.mount - /boot.
Jul 25 14:49:37 fedora systemd[1]: var.mount: Directory /var to mount over is not empty, mounting anyway.
Jul 25 14:49:37 fedora systemd[1]: Mounting var.mount - /var...
Jul 25 14:49:37 fedora kernel: EXT4-fs (dm-8): mounted filesystem 0b8645c2-60d4-4d3a-bcdb-eb2355965ac4 r/w with ordered data mode. Quota mode: none.
Jul 25 14:49:37 fedora systemd[1]: Mounted var.mount - /var.
Jul 25 14:49:37 fedora systemd[1]: Starting dracut-pre-mount.service - dracut pre-mount hook...
Jul 25 14:49:37 fedora systemd[1]: Finished dracut-pre-mount.service - dracut pre-mount hook.
Jul 25 14:49:37 fedora systemd[1]: Mounting home.mount - /home...
Jul 25 14:49:37 fedora systemd[1]: Mounting sysroot.mount - /sysroot...
Jul 25 14:49:37 fedora kernel: EXT4-fs (dm-12): mounted filesystem c8c018f9-a3fc-4cab-9f71-700be61ee322 r/w with ordered data mode. Quota mode: none.
Jul 25 14:49:37 fedora systemd[1]: Mounted home.mount - /home.
Jul 25 14:49:37 fedora kernel: EXT4-fs (dm-3): orphan cleanup on readonly fs
Jul 25 14:49:37 fedora kernel: EXT4-fs (dm-3): mounted filesystem 89564248-d7dc-4e98-ba3a-f5901ef1f756 ro with ordered data mode. Quota mode: none.
Jul 25 14:49:37 fedora systemd[1]: Mounted sysroot.mount - /sysroot.
Jul 25 14:49:37 fedora systemd[1]: Starting initrd-parse-etc.service - Mountpoints Configured in the Real Root...
Jul 25 14:49:37 fedora systemd[1]: Finished initrd-parse-etc.service - Mountpoints Configured in the Real Root.
Jul 25 14:49:37 fedora systemd[1]: dracut-mount.service - dracut mount hook was skipped because no trigger condition checks were met.
Jul 25 14:49:37 fedora systemd[1]: dracut-pre-mount.service: Deactivated successfully.
Jul 25 14:49:37 fedora systemd[1]: Stopped dracut-pre-mount.service - dracut pre-mount hook.
Jul 25 14:49:37 fedora systemd[1]: Starting plymouth-switch-root.service - Plymouth switch root service...
Jul 25 14:49:37 fedora systemd[1]: Finished plymouth-switch-root.service - Plymouth switch root service.
Jul 25 14:49:37 fedora systemd[1]: Reached target initrd-switch-root.target - Switch Root.
Jul 25 14:49:37 fedora systemd[1]: Starting initrd-switch-root.service - Switch Root...
Jul 25 14:49:38 fedora kernel: EXT4-fs (vda2): unmounting filesystem cbb2caf6-429f-4a6a-acc4-9a20564c3f58.
Jul 25 14:49:38 fedora kernel: EXT4-fs (dm-8): unmounting filesystem 0b8645c2-60d4-4d3a-bcdb-eb2355965ac4.
Jul 25 14:49:38 fedora kernel: EXT4-fs (dm-12): unmounting filesystem c8c018f9-a3fc-4cab-9f71-700be61ee322.
Jul 25 14:49:38 fedora systemd[1]: Stopped initrd-switch-root.service - Switch Root.
Jul 25 14:49:38 fedora systemd[1]: Set up automount proc-sys-fs-binfmt_misc.automount - Arbitrary Executable File Formats File System Automount Point.
Jul 25 14:49:38 fedora systemd[1]: Stopped target initrd-switch-root.target - Switch Root.
Jul 25 14:49:38 fedora systemd[1]: Mounting dev-hugepages.mount - Huge Pages File System...
Jul 25 14:49:38 fedora systemd[1]: Mounting dev-mqueue.mount - POSIX Message Queue File System...
Jul 25 14:49:38 fedora systemd[1]: Mounting sys-kernel-debug.mount - Kernel Debug File System...
Jul 25 14:49:38 fedora systemd[1]: Mounting sys-kernel-tracing.mount - Kernel Trace File System...
Jul 25 14:49:38 fedora systemd[1]: Stopped plymouth-switch-root.service - Plymouth switch root service.
Jul 25 14:49:38 fedora systemd[1]: Mounted dev-hugepages.mount - Huge Pages File System.
Jul 25 14:49:38 fedora systemd[1]: Mounted dev-mqueue.mount - POSIX Message Queue File System.
Jul 25 14:49:38 fedora systemd[1]: Mounted sys-kernel-debug.mount - Kernel Debug File System.
Jul 25 14:49:38 fedora systemd[1]: Mounted sys-kernel-tracing.mount - Kernel Trace File System.
Jul 25 14:49:38 fedora systemd[1]: Mounting sys-fs-fuse-connections.mount - FUSE Control File System...
Jul 25 14:49:38 fedora systemd[1]: Mounted sys-fs-fuse-connections.mount - FUSE Control File System.
Jul 25 14:49:38 fedora systemd[1]: Mounting home.mount - /home...
Jul 25 14:49:38 fedora systemd[1]: Mounting var.mount - /var...
Jul 25 14:49:38 fedora kernel: EXT4-fs (dm-12): mounted filesystem c8c018f9-a3fc-4cab-9f71-700be61ee322 r/w with ordered data mode. Quota mode: none.
Jul 25 14:49:38 fedora kernel: EXT4-fs (dm-8): mounted filesystem 0b8645c2-60d4-4d3a-bcdb-eb2355965ac4 r/w with ordered data mode. Quota mode: none.
Jul 25 14:49:38 fedora systemd[1]: Mounted home.mount - /home.
Jul 25 14:49:38 fedora systemd[1]: Mounted var.mount - /var.
Jul 25 14:49:38 fedora systemd[1]: tmp.mount: Directory /tmp to mount over is not empty, mounting anyway.
Jul 25 14:49:38 fedora systemd[1]: Mounting tmp.mount - Temporary Directory /tmp...
Jul 25 14:49:38 fedora systemd[1]: Mounted tmp.mount - Temporary Directory /tmp.
Jul 25 14:49:38 fedora systemd[1]: Unmounting home.mount - /home...
Jul 25 14:49:38 fedora kernel: EXT4-fs (dm-12): unmounting filesystem c8c018f9-a3fc-4cab-9f71-700be61ee322.
Jul 25 14:49:38 fedora kernel: EXT4-fs (vda2): mounted filesystem cbb2caf6-429f-4a6a-acc4-9a20564c3f58 r/w with ordered data mode. Quota mode: none.
Jul 25 14:49:38 fedora systemd[1]: Mounting boot.mount - /boot...
Jul 25 14:49:38 fedora systemd[1]: home.mount: Deactivated successfully.
Jul 25 14:49:38 fedora systemd[1]: Unmounted home.mount - /home.
Jul 25 14:49:38 fedora systemd[1]: Mounted boot.mount - /boot.
Jul 25 14:49:38 fedora systemd[1]: Unmounting var.mount - /var...
Jul 25 14:49:38 fedora umount[807]: umount: /var: target is busy.
Jul 25 14:49:38 fedora systemd[1]: var.mount: Mount process exited, code=exited, status=32/n/a
Jul 25 14:49:38 fedora systemd[1]: Failed unmounting var.mount - /var.
```

(the umount for /var fails since the target is busy but /home is unmounted at the end of booting).

**Comment by bmr-cymru** [comment]
Thanks! Does this have to do with the mount units being generated in the initramfs? I noticed in the logs that the command line mount units are interleaved with the mount of /sysroot (in the logs above dm-8 (/var) and dm-12 (/home) appear to be mounted *before* /sysroot??). This contrasts with normal boots where the auxiliary filesystems are all mounted after the switchroot.

I noticed that `add_mounts_from_cmdline()` always passes `initrd=false` which disables the `mount_in_initrd()` test in `parse_fstab_one()`. That means that all command line mount units will be mounted from the initrd regardless of the presence of `x-initrd.mount` in the mount options. That seems wrong.

Would it be better to have `add_mounts_from_cmdline()` also test `in_initrd()` and pass the `initrd` parameter to `parse_fstab_one()` accordingly? That way mounts that lack `x-initrd.mount`/are not `/usr` will be skipped over in the initramfs as with normal fstab processing.
